cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -506,10 +526,17 @@ public void upgradeApplication(HttpRequest request, HttpResponder responder,
                                  @QueryParam("artifactScope") Set<String> artifactScopes,
                                  @QueryParam("allowSnapshot") boolean allowSnapshot) throws Exception {
     ApplicationId appId = validateApplicationId(validateNamespace(namespaceId), appName);
-    Set<ArtifactScope> allowedArtifactScopes = getArtifactScopes(artifactScopes);
-    applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
-    ApplicationUpdateDetail updateDetail = new ApplicationUpdateDetail(appId);
-    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(updateDetail));
+    try {
+      String latestAppVersion = getLatestAppVersion(appId);
+      appId = validateApplicationVersionId(namespaceId, appName, latestAppVersion);
+      Set<ArtifactScope> allowedArtifactScopes = getArtifactScopes(artifactScopes);
+      applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
+      ApplicationUpdateDetail updateDetail = new ApplicationUpdateDetail(appId);
+      responder.sendJson(HttpResponseStatus.OK, GSON.toJson(updateDetail));
+    } catch (Exception e) {
+      LOG.error("Failure to upgrade application", e);
+      responder.sendJson(HttpResponseStatus.BAD_REQUEST, e.getMessage());
```

Comment:
Same as above - don't try to map everything to BAD_REQUEST, leave it to the general exception handler